pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -233,7 +233,9 @@ func (h *AppHandler) collectHealthStatus(ctx context.Context, wl *appfile.Worklo
 		if err := h.r.Client.Get(ctx, client.ObjectKey{Name: wl.Name, Namespace: namespace}, &configuration); err != nil {
 			return nil, false, errors.WithMessagef(err, "app=%s, comp=%s, check health error", appName, wl.Name)
 		}
-		if configuration.Status.Apply.State != terraformtypes.Available {
+		if appRev.Name != configuration.GetLabels()["app.oam.dev/appRevision"] ||
+			configuration.Status.Apply.State != terraformtypes.Available ||
+			configuration.Status.ObservedGeneration != configuration.Generation {
```

Comment:
Add an `&&` logic to check whether `configuration.Status.ObservedGeneration` is set.

Reply:
That good. here we also need to compare apprev

Reply:
What will happen if we  don't compare apprev?

Reply:
In kubebuilder, client.Get always gets data from the cache, so client.Get gets data from the cache immediately after applying object.

Ref: https://github.com/kubernetes-sigs/controller-runtime/issues/343